File: Examples/Simple Integration Obj-C/Source/OrderCreationViewController.m
//
//  OrderCreationViewController.m
//  Simple Integration Obj-C
//
//  Created by Johnny Peter on 29/08/19.
//  Copyright © 2019 Network International. All rights reserved.
//

#import "Simple_Integration_Obj_C-Swift.h"
#import "OrderCreationViewController.h"
#import "OrderRequest.h"
#import "OrderAmount.h"

@import NISdk;

@interface OrderCreationViewController ()
@property (nonatomic) int paymentAmount;
@property (nonatomic) NSArray<Product*> *purchasedItems;
@property (nonatomic) PaymentMethod paymentMethod;
@property (nonatomic) id<CardPaymentDelegate> cardPaymentDelegate;
@property (nonatomic) PKPaymentRequest *paymentRequest;
@end

@implementation OrderCreationViewController

- (instancetype)initWithPaymentAmount:(int)amount and:(id<CardPaymentDelegate>)cardPaymentDelegate using:(PaymentMethod)paymentMethod with:(NSArray<Product*> *)purchasedItems {
    self = [super init];
    if(self) {
        self.cardPaymentDelegate = cardPaymentDelegate;
        self.paymentAmount = amount;
        self.paymentMethod = paymentMethod;
        self.purchasedItems = purchasedItems;
    }
    
    if(self.paymentMethod == ApplePay) {
        self.paymentRequest = [[PKPaymentRequest alloc] init];
        NSString *merchantId = @"";
        NSAssert(merchantId.length > 0, @"You need to add your apple pay merchant ID above");
        self.paymentRequest.merchantIdentifier = merchantId;
        self.paymentRequest.countryCode = @"AE";
        self.paymentRequest.currencyCode = @"AED";
        self.paymentRequest.requiredShippingContactFields = [[NSSet alloc] initWithObjects: PKContactFieldPostalAddress, PKContactFieldEmailAddress, PKContactFieldPhoneNumber, nil];
        self.paymentRequest.merchantCapabilities = PKMerchantCapabilityDebit | PKMerchantCapabilityCredit | PKMerchantCapability3DS;
        self.paymentRequest.requiredBillingContactFields = [[NSSet alloc] initWithObjects: PKContactFieldPostalAddress, PKContactFieldName, nil];
        NSMutableArray<PKPaymentSummaryItem *> *summaryItems = [[NSMutableArray alloc] initWithCapacity: purchasedItems.count + 1];
        for (Product *item in self.purchasedItems) {
            [summaryItems addObject: [PKPaymentSummaryItem
                                      summaryItemWithLabel: item.name
                                      amount: [[NSDecimalNumber alloc] initWithFloat: item.amount]]];
        }
        [summaryItems addObject: [PKPaymentSummaryItem
                                  summaryItemWithLabel: @"NGenius merchant"
                                  amount: [[NSDecimalNumber alloc] initWithFloat: amount]]];
        self.paymentRequest.paymentSummaryItems = summaryItems;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.whiteColor;
    
    UILabel * authorizationLabel = [[UILabel alloc] init];
    [authorizationLabel setTextColor: UIColor.whiteColor];
    authorizationLabel.text = @"Creating Order...";
    
    UIActivityIndicatorView * spinner = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle: UIActivityIndicatorViewStyleGray];
    [spinner setHidden: true];
    [spinner setColor: UIColor.whiteColor];
    [spinner startAnimating];
    
    UIStackView *vStack = [[UIStackView alloc] initWithArrangedSubviews: [[NSArray alloc] initWithObjects: authorizationLabel, spinner, nil] ];
    [vStack setAxis: UILayoutConstraintAxisVertical];
    [vStack setSpacing: 0];
    [vStack setAlignment: UIStackViewAlignmentCenter];
    
    [self.view addSubview: vStack];
    [vStack anchorWithTop: nil leading: self.view.safeAreaLayoutGuide.leadingAnchor bottom: nil trailing: self.view.safeAreaLayoutGuide.trailingAnchor padding: UIEdgeInsetsZero size: CGSizeMake(0, 100)];
    
    [NSLayoutConstraint activateConstraints:@[
      [[vStack centerXAnchor] constraintEqualToAnchor: self.view.centerXAnchor],
      [[vStack centerYAnchor] constraintEqualToAnchor: self.view.centerYAnchor]]];
    
    [self createOrder];
   
}

- (void)dismissVC {
    dispatch_async(dispatch_get_main_queue(), ^(void){
        [self dismissViewControllerAnimated: false completion: nil];
    });
}

- (void)displayErrorAndClose:(NSError*)error {
    NSString *errorTitle = @"";
    if(error != nil) {
        NSDictionary *userInfo = [error userInfo];
        errorTitle = (NSString*)[userInfo objectForKey: @"NSLocalizedDescription"];
        if(errorTitle == nil) {
            errorTitle = @"Unknown Error";
        }
    }
    
    dispatch_async(dispatch_get_main_queue(), ^(void){
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:errorTitle message:@"" preferredStyle: UIAlertControllerStyleAlert];
        [alert addAction: [UIAlertAction actionWithTitle: @"Ok" style:UIAlertActionStyleDefault handler: ^(UIAlertAction * action){
            [self dismissVC];
        }]];
        [self presentViewController:alert animated:true completion:nil];
        
    });
}

- (void)createOrder {
    OrderRequest *orderRequest = [[OrderRequest alloc] initWithAction: @"SALE" andAmount: [[OrderAmount alloc] initWitCurrencyCode: @"AED" andValue: self.paymentAmount * 100]];
    
    NSDictionary *requestHeaders = [[NSDictionary alloc] initWithObjectsAndKeys: @"application/json", @"Content-Type", nil];
    
    NSData *orderRequestData = [NSJSONSerialization dataWithJSONObject:[orderRequest dictionary]  options: NSJSONWritingSortedKeys error: nil];
    
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL: [NSURL URLWithString: @"http://localhost:3000/api/createOrder"] cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval: 10.0];
    
    [request setHTTPMethod: @"POST"];
    [request setAllHTTPHeaderFields: requestHeaders];
    [request setHTTPBody:orderRequestData];
    
    NSURLSession *session = [NSURLSession sharedSession];
    
    @try {
        __weak OrderCreationViewController *weakSelf = self;
        NSURLSessionDataTask *dataTask = [session dataTaskWithRequest: request completionHandler: ^(NSData *data, NSURLResponse *response, NSError *error) {
            if(error != nil) {
                [weakSelf displayErrorAndClose: error];
                return;
            }
            if(data != nil) {
                OrderResponse *orderResponse = [OrderResponse decodeFromData: data error: nil];
                dispatch_async(dispatch_get_main_queue(), ^(void){
                    [weakSelf dismissViewControllerAnimated: NO completion: ^() {
                        NISdk *sdkInstance = [NISdk sharedInstance];
                        if(weakSelf.paymentMethod == Card) {
                            [sdkInstance showCardPaymentViewWithCardPaymentDelegate: weakSelf.cardPaymentDelegate
                                                                         overParent:(UIViewController *)weakSelf.cardPaymentDelegate
                                                                                for: orderResponse];
                        } else {
                            [sdkInstance initiateApplePayWithApplePayDelegate: (id<ApplePayDelegate>)weakSelf.cardPaymentDelegate
                                                          cardPaymentDelegate: weakSelf.cardPaymentDelegate
                                                                   overParent: (UIViewController *)weakSelf.cardPaymentDelegate
                                                                          for: orderResponse
                                                                         with: self.paymentRequest];
                        }
                    }];
                });
            }
        }];
        [dataTask resume];
    } @catch (NSException *exception) {
        __weak OrderCreationViewController *weakSelf = self;
        NSMutableDictionary * info = [NSMutableDictionary dictionary];
        [info setValue:exception.name forKey:@"ExceptionName"];
        [info setValue:exception.reason forKey:@"ExceptionReason"];
        [info setValue:exception.callStackReturnAddresses forKey:@"ExceptionCallStackReturnAddresses"];
        [info setValue:exception.callStackSymbols forKey:@"ExceptionCallStackSymbols"];
        [info setValue:exception.userInfo forKey:@"ExceptionUserInfo"];
        NSError *error = [[NSError alloc] initWithDomain: NSURLErrorDomain code: 0 userInfo:info];
        [weakSelf displayErrorAndClose: error];
    }
}

@end

File: Examples/Simple Integration/Source/OrderCreationViewController.swift
//
//  OrderCreationViewController.swift
//  Simple Integration
//
//  Created by Johnny Peter on 23/08/19.
//  Copyright © 2019 Network International. All rights reserved.
//

import Foundation
import UIKit
import NISdk
import PassKit

/// Creates an order on the N-Genius gateway and passes the response back to the store front.
///
/// Key details:
///   - The payment amount must be multiplied by 100 (minor units) before sending to the API.
///   - After a successful order, the SDK language is set via `NISdk.sharedInstance.setSDKLanguage`.
///   - On failure, an error alert is shown and the view controller is dismissed.
class OrderCreationViewController: UIViewController {
    let paymentAmount: Double
    let storeFrontDelegate: StoreFrontDelegate
    let purchasedItems: [Product]
    var savedCard: SavedCard?
    var onOrderCreated: ((OrderResponse) -> Void)?

    let apiService = ApiService()

    init(paymentAmount: Double,
         storeFrontDelegate: StoreFrontDelegate,
         with purchasedItems: [Product],
         savedCard: SavedCard? = nil,
         onOrderCreated: @escaping (OrderResponse) -> Void) {
        self.paymentAmount = paymentAmount
        self.storeFrontDelegate = storeFrontDelegate
        self.purchasedItems = purchasedItems
        self.savedCard = savedCard
        self.onOrderCreated = onOrderCreated
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func dismissVC() {
        DispatchQueue.main.async {
            self.dismiss(animated: false, completion: nil)
        }
    }
    
    func displayErrorAndClose(error: Error?) {
        var errorTitle = "Unknown Error"
        var errorMessage = ""
        if let error = error {
            let nsError = error as NSError
            let userInfo = nsError.userInfo
            errorTitle = userInfo["NSLocalizedDescription"] as? String
                ?? nsError.localizedDescription
            errorMessage = "Domain: \(nsError.domain), Code: \(nsError.code)"
        }

        DispatchQueue.main.async {
            let alert = UIAlertController(title: errorTitle, message: errorMessage, preferredStyle: UIAlertController.Style.alert)
            alert.addAction(UIAlertAction(title: "Ok", style: UIAlertAction.Style.default, handler: { [weak self] _ in self?.dismissVC() }))
            self.present(alert, animated: true, completion: nil)
        }
    }
    
    func createOrder(savedCard: SavedCard? = nil) {
        // Multiply amount always by 100 while creating an order
        let merchantAttributes = Environment.getMerchantAttributes()
        let attributeDictionary: [String: String]? = if (merchantAttributes.isEmpty) {
            nil
        } else {
            merchantAttributes.reduce(into: [String: String]()) { dict, attribute in
                dict[attribute.key] = attribute.value
            }
        }
        let currencyCode = Environment.getCurrency()
        var orderRequest = OrderRequest(action: Environment.getOrderAction(),
                                        amount: OrderAmount(currencyCode: currencyCode, value: paymentAmount * 100),
                                        language: Environment.getLanguage(),
                                        merchantAttributes: attributeDictionary,
                                        savedCard: savedCard)

        // add required parameters for order type
        let orderType = Environment.getOrderType()
        switch orderType {
            case "INSTALLMENT":
                orderRequest.installmentDetails = InstallmentDetails(numberOfTenure: 2)
                orderRequest.type = "INSTALLMENT"
                orderRequest.frequency = "MONTHLY"
            case "UNSCHEDULED":
                orderRequest.type = "UNSCHEDULED"
            case "RECURRING":
                orderRequest.type = "RECURRING"
                orderRequest.frequency = "MONTHLY"
                orderRequest.recurringDetails = RecurringDetails(numberOfTenure: 10, recurringType: "FIXED")
            default:
                break
        }
        
        apiService.createOrder(orderData: orderRequest) { result in
            switch result {
            case .success(let orderResponse):
                self.storeFrontDelegate.updateOrderId(orderId: orderResponse.reference ?? "")
                NISdk.sharedInstance.setSDKLanguage(language: Environment.getLanguage())
                DispatchQueue.main.async {
                    self.dismiss(animated: false, completion: { [weak self] in
                        self?.onOrderCreated?(orderResponse)
                    })
                }
            case .failure(let error):
                self.displayErrorAndClose(error: error)
            }
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let authorizationLabel = UILabel()
        authorizationLabel.textColor = .white
        authorizationLabel.text = "Creating Order..."
        let spinner = UIActivityIndicatorView(style: .medium)
        spinner.isHidden = false
        spinner.color = .white
        spinner.startAnimating()
        
        let vStack = UIStackView(arrangedSubviews: [authorizationLabel, spinner])
        vStack.axis = .vertical
        vStack.spacing = 0
        vStack.alignment = .center
        
        view.addSubview(vStack)
        vStack.anchor(top: nil,
                      leading: view.safeAreaLayoutGuide.leadingAnchor,
                      bottom: nil,
                      trailing: view.safeAreaLayoutGuide.trailingAnchor,
                      padding: .zero,
                      size: CGSize(width: 0, height: 100))
        
        vStack.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        vStack.centerYAnchor.constraint(equalTo: view.centerYAnchor).isActive = true
        self.createOrder(savedCard: savedCard)
    }
}
